doc/proposals/ansible-operator-webhooks.md

```diff
@@ -0,0 +1,81 @@
+
+## Non-Goals
+- Handle xWebhookConfiguration objects or associated Services
+- Handle cert management
+
+## Solution
+Add a webhooks section to the `watches.yaml`, that allows the user to specify lists of validating/mutating/conversion 
+webhooks that should be created. 
+
+The user should only need to provide a playbook or a role, and the url path for the webhook. 
+
+The rest of the webhook configuration, ie the creation of the Service that points to the webhook server, the Webhook 
+Configuration resources that lay out the rules and point to the proper Service, and the management of the certificates 
+for the webhook server, will be the responsiblity of the user. 
+
+The webhook configuration will look roughly as follows, with a big asterisk on the conversion webhooks which need some
+further investigation.
+
+```yaml
+---
+- version: v1alpha1
+  group: apps.example.com
+  kind: MyApp
+  role: /opt/ansible/roles/myapp
+  webhooks:
+    validating:
+    - playbook: /opt/ansible/validate.yml
+      path: /validate
+    mutating:
+    - role: /opt/ansible/roles/mutate-myapp
+      path: /mutate
+    conversion:
+    - playbook: /opt/ansible/conversion.yml
+      path: /convert
+```
+
+The justification for allowing multiple webhooks per type is that webhooks can select objects by a variety of
+attributes, so you could have a different webhook which is selected based on `metadata.labels`.
+
+Based on the webhook configuration, during initialization we will create a webhook server and add it to manager,
+and then register each of the webhooks to the webhook server, at the given path. We will create handlers that
+respond to the requests by running the associated Ansible roles or playbooks.
+
+- For validating webhooks, playbook success => Allowed, and playbook failure => Denied. We will report the
+    failureMessages in the denial response.
+- For mutating webhooks, we will create a custom module that will be used to report back the final desired
+    object. We may need to write additional logic here (either in the module or on the Go-side) that generates
+    a proper patch object from this.
+- Conversion webhooks should work roughly the same as mutating webhooks.
+
+## Discussion / Further Investigation
+- For non-trivial playbooks, will the performance be acceptable?
+- How much webhook configuration should we support?
+- Should we support side-effects?
+    Side-effects will require adding reconciliation mechanisms to the webhook,
+    and would likely impact the performance even more extremely. My instinct is to not support it as it adds
+    a lot of complexity and seems dangerous, but then we're obviously restricting the power of the webhooks.
+    If we don't allow side-effects, we could enforce that by making the proxy reject requests from the webhooks
+    that do anything but `GET`.
+- Need to determine exactly what information we should send to the webhooks, and how it will be structured.
```

Comment:
It might be nice to see what is sent to ansible runner for these in the proposal?

Reply:
I should clarify here, I mean how much of the request object, or should we just pass the whole request object in.